gridded_etl_tools/utils/publish.py

```diff
@@ -793,35 +793,147 @@ def check_written_value(
         ----------
         orig_ds
-            The original dataset, unformatted
+            The original dataset, formatted minimally to allow comparison with the production dataset
         """
-        # Randomly select an original dataset
-        raw_ds, orig_file_path = self.binary_search_for_file(coords[self.time_dim], time_dim=self.time_dim)
+        # Apply any custom filters to the input files. Returns all input files if no custom filter defined.
+        possible_files = self.custom_file_filter(list(self.input_files()))
+
+        # Select an original dataset containing the same time dimension value as the production dataset
+        # Orig_file_path is unneeded but useful for debugging
+        orig_ds, orig_file_path = self.binary_search_for_file(
+            coords[self.time_dim], time_dim=self.time_dim, possible_files=possible_files
+        )
 
         # If a forecast dataset then search again, this time for the correct step
         if "step" in coords:
-            frt_string = self.numpydate_to_py(np.atleast_1d(raw_ds[self.time_dim])[0]).date().isoformat()
-            date_filtered_original_files = [fil for fil in list(self.input_files()) if frt_string in str(fil)]
-            raw_ds, orig_file_path = self.binary_search_for_file(
-                coords["step"], time_dim="step", possible_files=date_filtered_original_files
+            time_filtered_original_files = self.filter_files_by_time(original_files=possible_files, raw_ds=orig_ds)
+            orig_ds, orig_file_path = self.binary_search_for_file(
+                coords["step"], time_dim="step", possible_files=time_filtered_original_files
+            )
+
+        # If an ensemble dataset then search again, this time for the correct ensemble number
+        if "ensemble" in coords:
+            step_filtered_original_files = self.filter_files_by_step(
+                original_files=time_filtered_original_files, raw_ds=orig_ds
+            )
+            orig_ds, orig_file_path = self.binary_search_for_file(
+                coords["ensemble"], time_dim="ensemble", possible_files=step_filtered_original_files
+            )
+
+        # If a hindcast dataset then search again, this time for the correct forecast_reference_offset
+        if "forecast_reference_offset" in coords:
+            ensemble_filtered_original_files = self.filter_files_by_ensemble(
+                original_files=step_filtered_original_files, raw_ds=orig_ds
+            )
+            orig_ds, orig_file_path = self.binary_search_for_file(
+                coords["forecast_reference_offset"],
+                time_dim="forecast_reference_offset",
+                possible_files=ensemble_filtered_original_files,
             )
 
-        # Reformat the dataset such that it can be selected from equivalently to the prod dataset
-        orig_ds = self.reformat_orig_ds(raw_ds, orig_file_path)
         return orig_ds
 
+    def custom_file_filter(self, original_files: tuple[str]) -> tuple[str]:
```

Comment:
I guess this is somewhat subjective, but it seems redundant to call this a custom filter. If it's the object's filter function, it's not really custom, even if it's a modification of the parent's function. My expectation would be a custom filter would be a function that is passed in as an argument.